src/lazy/text/buffer.rs

```diff
@@ -0,0 +1,732 @@
+#[derive(PartialEq, Clone, Copy)]
+pub(crate) struct TextBufferView<'a> {
+    // `data` is a slice of remaining data in the larger input stream.
+    // `offset` is the absolute position in the overall input stream where that slice begins.
+    //
+    // input: 00 01 02 03 04 05 06 07 08 09
+    //                          └────┬────┘
+    //                          data: &[u8]
+    //                          offset: 6
+    data: &'a [u8],
+    offset: usize,
+}
+
+pub(crate) type ParseResult<'a, T> = IonResult<(T, TextBufferView<'a>)>;
+
+impl<'data> TextBufferView<'data> {
+    /// Constructs a new `TextBufferView` that wraps `data`, setting the view's `offset` to zero.
+    #[inline]
+    pub fn new(data: &[u8]) -> TextBufferView {
+        Self::new_with_offset(data, 0)
+    }
+
+    /// Constructs a new `TextBufferView` that wraps `data`, setting the view's `offset` to the
+    /// specified value. This is useful when `data` is a slice from the middle of a larger stream.
+    /// Note that `offset` is the index of the larger stream at which `data` begins and not an
+    /// offset _into_ `data`.
+    pub fn new_with_offset(data: &[u8], offset: usize) -> TextBufferView {
+        TextBufferView { data, offset }
+    }
+
+    /// Returns a subslice of the [`TextBufferView`] that starts at `offset` and continues for
+    /// `length` bytes.
+    ///
+    /// Note that `offset` is relative to the beginning of the buffer, not the beginning of the
+    /// larger stream of which the buffer is a piece.
+    pub fn slice(&self, offset: usize, length: usize) -> TextBufferView<'data> {
+        TextBufferView {
+            data: &self.data[offset..offset + length],
+            offset: self.offset + offset,
+        }
+    }
+
+    /// Returns a subslice of the [`TextBufferView`] that starts at `offset` and continues
+    /// to the end.
+    ///
+    /// Note that `offset` is relative to the beginning of the buffer, not the beginning of the
+    /// larger stream of which the buffer is a piece.
+    pub fn slice_to_end(&self, offset: usize) -> TextBufferView<'data> {
+        TextBufferView {
+            data: &self.data[offset..],
+            offset: self.offset + offset,
+        }
+    }
+
+    /// Returns a slice containing all of the buffer's bytes.
+    pub fn bytes(&self) -> &[u8] {
+        self.data
+    }
+
+    /// Returns the number of bytes between the start of the original input byte array and the
+    /// subslice of that byte array that this `TextBufferView` represents.
+    pub fn offset(&self) -> usize {
+        self.offset
+    }
+
+    /// Returns the number of bytes in the buffer.
+    pub fn len(&self) -> usize {
+        self.data.len()
+    }
+
+    /// Returns `true` if there are no bytes in the buffer. Otherwise, returns `false`.
+    pub fn is_empty(&self) -> bool {
+        self.data.is_empty()
+    }
+
+    pub fn match_whitespace(self) -> IonMatchResult<'data> {
```

Comment:
🗺️ Note that only a handful of the `match_` parsing methods are `pub`. Most are helper-parsers.